server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/ReindexingRoutes.java

```diff
@@ -0,0 +1,303 @@
+import io.swagger.annotations.ApiOperation;
+import io.swagger.annotations.ApiResponse;
+import io.swagger.annotations.ApiResponses;
+import spark.Request;
+import spark.Response;
+import spark.Service;
+
+@Api(tags = "ReIndexing")
+@Path("/mailboxIndex")
+@Produces("application/json")
+public class ReindexingRoutes implements Routes {
+    private static final String BASE_PATH = "/mailboxIndex";
+    private static final String USER_PARAM = ":user";
+    private static final String MAILBOX_PARAM = ":mailbox";
+    private static final String UID_PARAM = ":uid";
+    private static final String USER_PATH = BASE_PATH + "/user/" + USER_PARAM;
+    private static final String MAILBOX_PATH = USER_PATH + "/mailbox/" + MAILBOX_PARAM;
+    private static final String MESSAGE_PATH = MAILBOX_PATH + "/uid/" + UID_PARAM;
+
+    private final TaskManager taskManager;
+    private final MailboxManager mailboxManager;
+    private final MailboxId.Factory mailboxIdFactory;
+    private final ReIndexer reIndexer;
+    private final JsonTransformer jsonTransformer;
+
+    @Inject
+    public ReindexingRoutes(TaskManager taskManager, MailboxManager mailboxManager, MailboxId.Factory mailboxIdFactory, ReIndexer reIndexer, JsonTransformer jsonTransformer) {
+        this.taskManager = taskManager;
+        this.mailboxManager = mailboxManager;
+        this.mailboxIdFactory = mailboxIdFactory;
+        this.reIndexer = reIndexer;
+        this.jsonTransformer = jsonTransformer;
+    }
+
+    @Override
+    public String getBasePath() {
+        return BASE_PATH;
+    }
+
+    @Override
+    public void define(Service service) {
+        service.post(BASE_PATH, this::reIndexAll, jsonTransformer);
+        service.post(USER_PATH, this::reIndexUser, jsonTransformer);
+        service.post(MAILBOX_PATH, this::reIndexMailbox, jsonTransformer);
+        service.post(MESSAGE_PATH, this::reIndexMessage, jsonTransformer);
+    }
+
+    @POST
+    @Path("/")
+    @ApiOperation(value = "Re-indexes all the mails on this server")
+    @ApiImplicitParams({
+        @ApiImplicitParam(
+            required = true,
+            name = "task",
+            paramType = "query parameter",
+            dataType = "String",
+            defaultValue = "none",
+            example = "?task=reIndex",
+            value = "Compulsory. Only supported value is `reIndex`")
+    })
+    @ApiResponses(value = {
+        @ApiResponse(code = HttpStatus.CREATED_201, message = "Task is created", response = TaskIdDto.class),
+        @ApiResponse(code = HttpStatus.INTERNAL_SERVER_ERROR_500, message = "Internal server error - Something went bad on the server side."),
+        @ApiResponse(code = HttpStatus.BAD_REQUEST_400, message = "Bad request - details in the returned error message")
+    })
+    private TaskIdDto reIndexAll(Request request, Response response) {
+        return wrap(request, response,
+            () -> reIndexer.reIndex());
+    }
+
+    @POST
+    @Path("/user/{user}")
```

Comment:
s/user/users : when you expect that something contains a resource (a given user) the path at the previous level is supposed to be a plural (and can allow to list resources)